cmd/fs-v1.go

```diff
@@ -72,15 +73,124 @@ func initMetaVolumeFS(fsPath, fsUUID string) error {
 
 }
 
+// Migrate FS object is a place holder code for all
+// FS format migrations.
+func migrateFSObject(fsPath, fsUUID string) (err error) {
+	return migrateFSFormatV1ToV2(fsPath, fsUUID)
+}
+
+// Attempts to migrate old object metadata files to newer format
+//
+// i.e
+//    -------------------------------------------------------
+//    .minio.sys/buckets/<bucket_name>/<object_path>/fs.json - V1
+//    -------------------------------------------------------
+//    .minio.sys/buckets/<bucket_name>/objects/<object_path>/fs.json - V2
+//    -------------------------------------------------------
+//
+func migrateFSFormatV1ToV2(fsPath, fsUUID string) (err error) {
+	metaBucket := pathJoin(fsPath, minioMetaBucket, bucketConfigPrefix)
+	migratePlaceHolder := "." + fsUUID
+
+	// Locked for writing ".migrate" file under the buckets folder.
+	var lk *lock.LockedFile
+	lk, err = lock.LockedOpenFile(preparePath(pathJoin(metaBucket, migratePlaceHolder)),
+		os.O_CREATE|os.O_WRONLY, 0600)
+	if err != nil {
+		return err
+	}
+	defer lk.Close()
+
+	_, err = lk.Write([]byte(""))
+	if err != nil {
+		return err
+	}
+
+	// Writing message here is important for old servers.
+	log.Println("PLEASE DO NOT STOP THE SERVER")
+	log.Printf("Migrating old bucket metadata format at \"%s\" \"1\" to newer format \"2\"... ", fsPath)
+
+	defer func() {
+		if err == nil {
+			log.Println("SUCCESS.")
+		} else {
+			log.Printf("FAILED with %s.\n", err)
+		}
+	}()
+
+	var buckets []string
+	buckets, err = fsReaddirBuckets(fsPath)
```

Comment:
> We are constructing it internally @vadmeste

You mean we are going to create all buckets inside .minio.sys/buckets?

Anyway, it will fail later as I said in fsRenameFile(), I just tested it. pathJoin(metaBucket, bucket) doesn't always exist and it returns an error at that stage.